Add release notes for v2.2.1 (#1877)
hyperledger/fabric: release_notes/v2.2.1.md

If using a separate cluster listener port,
configure the orderer `General.Cluster.TLSHandshakeTimeShift` property.

**FAB-18205: peer CLI certificate expiration - Permit peer CLI to communicate with orderers with expired TLS certs**

The change allows peer CLI to communicate with ordering service nodes with expired TLS certificates
by setting the `--tlsHandshakeTimeShift` flag to a desired backwards timeshift.
The change applies to the `peer channel fetch` and `peer channel update` commands to allow
fetching configuration blocks and updating the channel configuration for orderers with expired TLS certificates.

**FAB-18171: orderer certificate expiration - Disregard certificate validity period in intra-orderer communication**

This change makes the orderer cluster authentication infrastructure
disregard validity periods when comparing certificates, and only regard public keys.
With this change, an expiring Raft TLS certificate can be replaced
with a new certificate that has the same public key, without issuing channel configuration updates.

**FAB-18188: peer and orderer certificate expiration - Log expiration date upon startup**

> The future enrollment, tls server, and tls client certificate expiration dates are now logged upon peer and orderer startup.

Reviewer: What does "the future" mean? 
Maybe: "The enrollment, tls server and client certificate expiration dates are now logged..." 
why do we need "the future" here? It's not necessarily in the future, it can be also in the past

Author: Done

**peer and orderer HSM - Drain session pool before creating new sessions**

If a bad session was retrieved from the session pool, a new session
would be immediately created instead of attempting to retrieve another
session from the pool.
This change attempts to uses pooled sessions before creating new ones.

**peer and orderer HSM - Add object handle cache to PKCS#11 provider**

With this change, object handles are cached in the PKCS#11 implementation.
In support of this change, in addition to pooling idle sessions, the
provider tracks active sessions. If some condition occurs that results
in all sessions being closed, cached object handles are no longer valid
so the handle cache is purged.

**FAB-18208: Do not sign gossip message if membership is empty**

This change suppresses the signing of gossip messages if the message will not get

> sent regardless due to an empty gossip membership. The change improves performance,

Reviewer: Doesn't improve performance, it reduces CPU consumption and stress on HSM

Author: Done

If using a separate cluster listener port,
configure the orderer.yaml `General.Cluster.TLSHandshakeTimeShift` property.

**FAB-18205: orderer certificate expiration - Permit peer CLI to communicate with orderers with expired TLS certificates**

The change allows peer CLI to communicate with ordering service nodes with expired TLS certificates
by setting the `--tlsHandshakeTimeShift` flag to a desired backwards timeshift.
The change applies to the `peer channel fetch` and `peer channel update` commands to allow
fetching configuration blocks and updating the channel configuration for orderers with expired TLS certificates.

**FAB-18171: orderer certificate expiration - Disregard certificate validity period in intra-orderer communication**

This change makes the orderer cluster authentication infrastructure
disregard validity periods when comparing certificates, and only regard public keys.
With this change, an expiring Raft TLS certificate can be replaced
with a new certificate that has the same public key, without requiring channel configuration updates.

**FAB-18188: peer and orderer certificate expiration - Log expiration date upon startup**

> The enrollment, tls server, and tls client certificate expiration dates are now logged upon peer and orderer startup.

Reviewer 2: ```suggestion
The enrollment, TLS server, and TLS client certificate expiration dates are now logged upon peer and orderer startup.
```

> **peer and orderer HSM - Drain session pool before creating new sessions**

Reviewer 2: Should we say ` PKCS#11` rather than HSM?

> **peer and orderer HSM - Drain session pool before creating new sessions**

Reviewer 2: ```suggestion
**peer and orderer PKCS#11 - Drain session pool before creating new sessions**
```

> If a bad hardware security module (HSM) session was retrieved from the session pool,

Reviewer 2: ```suggestion
If an invalid PKCS#11 session was acquired from the session pool,
```

> a new session would be immediately created instead of attempting to retrieve another

Reviewer 2: ```suggestion
a new session would be immediately created instead of attempting to acquire another
```

session from the pool.
This change attempts to uses pooled sessions before creating new ones.

> **peer and orderer HSM - Add object handle cache to PKCS#11 provider**

Reviewer 2: ```suggestion
**peer and orderer PKCS#11 - Add object handle cache**
```

With this change, object handles are cached in the PKCS#11 implementation.
In support of this change, in addition to pooling idle sessions, the
provider tracks active sessions. If some condition occurs that results
in all sessions being closed, cached object handles are no longer valid
so the handle cache is purged.

**FAB-18208: Do not sign gossip message if membership is empty**

This change suppresses the signing of gossip messages if the message will not get
sent regardless due to an empty gossip membership. The change reduces CPU consumption
and eliminates unnecessary calls to an HSM.

> **peer and orderer HSM - Add log message if a key cannot be found in the HSM**

Reviewer 2: ```suggestion
**peer and orderer PKCS#11 - Add log message if a key cannot be found**
```

**FAB-18194: peer - Fix service discovery for legacy installed chaincodes**

The service discovery endorsers query may fail with errors
"failed constructing descriptor for chaincodes" and
"required chaincodes are not installed on sufficient peers", even
when a chaincode is installed on a sufficient number of peers.
The peer initialization has been fixed so that chaincodes installed with the legacy chaincode
lifecycle are correctly recognized by service discovery.

**FAB-18191: peer - Remove contents of leveldb directory instead of the directory when dropping databases**

The peer `upgrade-dbs`, `rebuild-dbs`, `reset`, and `rollback` commands each drop a peer's
leveldb databases so that they can be rebuilt upon the next peer startup.
The commands now remove all contents of the leveldb directories, rather than dropping the
directory itself. This fix is required if mounting one of the leveldb directories in addition
to the overall peer data directory as specified in core.yaml `peer.fileSystemPath`.

**FAB-15810: peer - Deprioritize fetching of missing private data**

The private data reconciler attempts to retrieve missing private data from other

> peers in a channel that belong to the same private data collection, based on the interval

Reviewer 2: ```suggestion
peers in a channel that belong to the same private data collection based on the interval
```

configured in core.yaml `peer.gossip.pvtData.reconcileSleepInterval`
(by default every one minute). If private data cannot be reconciled, it will be attempted
every interval and may block the reconciliation of other private data. This fix places
private data that can't be reconciled in a depriorized queue so that other private data
can be reconciled. The depriorized entries will be re-attempted less often based on the
interval configured in core.yaml `ledger.pvtdataStore.deprioritizedDataReconcilerInterval`
(by default every 60 minutes).

**orderer - Check suspect info once per suspect interval when using Raft**

> The raft-based ordering service node was checking to see if it was evicted too often.

Reviewer 2: ```suggestion
The Raft-based ordering service node was checking to see if it was evicted too often.
```

This fix ensures that the ordering service node only checks once per suspect interval,

> every 10 minutes by default.

Reviewer 2: ```suggestion
which is every 10 minutes by default.
```
